server/api_archive_integration_test.go

```diff
@@ -0,0 +1,255 @@
+		go func() {
+			defer wg.Done()
+
+			// Randomly choose a point within each segment
+			durs := make([]float64, apiArchiveLimit)
+			for j := 0; j < apiArchiveLimit; j++ {
+				durs[j] = (float64(j) + rand.Float64()) * segment.Seconds()
+			}
+
+			slices.Sort(durs) // Make sure the times are in increasing order
+
+			cuttab := make([]time.Time, apiArchiveLimit)
+			for j := 0; j < apiArchiveLimit; j++ {
+				cuttab[j] = start.Add(time.Duration(durs[j] * float64(time.Second)))
+			}
+
+			for j := 0; j < apiArchiveLimit; j++ {
+				// Sleep until the next cut time.
+				time.Sleep(time.Until(cuttab[j]))
+
+				resp, err := http.Get(fmt.Sprintf("http://localhost:%v/api/v1/archive", gcas.httpPort))
+				if err != nil {
+					t.Fatal(err)
+				}
+				switch resp.StatusCode {
+				case http.StatusOK:
+					allowed.Add(1)
+				case http.StatusTooManyRequests:
+					denied.Add(1)
+				default:
+					t.Fatalf("Archive API call returned an invalid response %v", resp)
+				}
+			}
+		}()
+	}
+	wg.Wait()
+
+	testDur := time.Since(start)
+
+	if testDur > time.Duration(3)*apiArchiveRate {
+		t.Errorf("duration %v exceeded %v", testDur, time.Duration(3)*apiArchiveRate)
+	}
+
+	if allowed.Load()+denied.Load() != int32(threads*apiArchiveLimit) {
+		t.Errorf("incorrect requests %v, expected %v", allowed.Load()+denied.Load(), threads*apiArchiveLimit)
+	}
+
+	// Calculate the number of rate limit intervals.
+	periods := int(testDur / apiArchiveRate)
+	if testDur%apiArchiveRate != 0 {
+		periods++
+	}
+
+	if int(allowed.Load()) > periods*apiArchiveLimit {
+		t.Errorf("%v allowed, expected %v", allowed.Load(), periods*apiArchiveLimit)
+	}
+}
+
+// Create a server environment with populated archive files.
+// Returns a running GCA server, and the test environment directory name.
+func ServerWithArchiveFiles(name string) (*GCAServer, string, error) {
+	gcas, dirname, _, gcaPrivKey, err := SetupTestEnvironment(name)
+	if err != nil {
+		return nil, "", err
+	}
+	// This test is going to be messing with time, therefore defer a reset
+	// of the time.
+	defer glow.SetCurrentTimeslot(0)
+
+	// Generate a keypair for a device.
+	authPub, authPriv := glow.GenerateKeyPair()
+	auth := glow.EquipmentAuthorization{ShortID: 0, PublicKey: authPub}
+	sb := auth.SigningBytes()
+	sig := glow.Sign(sb, gcaPrivKey)
+	auth.Signature = sig
+	_, err = gcas.saveEquipment(auth)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Submit reports for slots 0, 2, and 4.
+	err = gcas.sendEquipmentReportSpecific(auth, authPriv, 0, 50)
+	if err != nil {
+		return nil, "", err
+	}
+	err = gcas.sendEquipmentReportSpecific(auth, authPriv, 2, 50)
+	if err != nil {
+		return nil, "", err
+	}
+	err = gcas.sendEquipmentReportSpecific(auth, authPriv, 4, 50)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Submit reports for slots 4031, 4030, and 4028. For these reports to
+	// be accepted, time must be shifted. This will also trigger a report
+	// migration.
+	glow.SetCurrentTimeslot(4000)
```

Comment:
When this function is called, the timeslot needs to be reset back to 0 at the end of the test. I believe you can see examples of how this is done in the client integration tests.

That also means this helper function needs to be clearly documented that it makes changes to the current timeslot, because anyone who calls it needs to know that they need to `defer` a reset to the timeslot otherwise it'll mess up other tests.

Reply:
Line 206 does have a reset:
	`// This test is going to be messing with time, therefore defer a reset
	// of the time.
	defer glow.SetCurrentTimeslot(0)`